### PR TITLE
server: fix for inactive service offering for VM

### LINKED ISSUE #3315: Set an inactive service offering over API
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.11.2
~~~

##### CONFIGURATION

ANY

##### SUMMARY
We discovered something in our cloud. It is possible to set up an inactive service offering for a VM over API. This is not possible over UI as the service offer disappears after deleting it. However, over API we are still able to set up the service offer which was "deleted". 

##### STEPS TO REPRODUCE

~~~
1. Create a new service offer (compute offering). 
2. Note the ID of this new service offer and delete it
3. shut down any test VM and set the just deleted service offer over API:
change serviceforvirtualmachine id=d9a82873-XXXX-XXXX-XXXX-81cbf1ae6133 serviceofferingid=8f352ab6-XXXX-XXXX-XXXX-91e393186862
4. You can start the VM again and use the service offering which is set as "Inactive"
~~~

##### EXPECTED RESULTS

~~~
API Call should fail with the service offering inactive
~~~

##### ACTUAL RESULTS

~~~
API call works, the inactive service offering is used
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
## Description
Fixes #3315
Currently, the code was allowed to change service offering for VM to a deleted or inactive service offering. Added check for it to throw an exception.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
Using cmk

### CODE REVIEW

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -965,6 +965,9 @@ public UserVm upgradeVirtualMachine(UpgradeVMCmd cmd) throws ResourceAllocationE
         // Check resource limits for CPU and Memory.
         Map<String, String> customParameters = cmd.getDetails();
         ServiceOfferingVO newServiceOffering = _offeringDao.findById(svcOffId);
+        if (newServiceOffering.getState() == DiskOffering.State.Inactive) {
+            throw new InvalidParameterValueException(String.format("Unable to upgrade virtual machine %s with an inactive service offering %s", vmInstance.toString(), newServiceOffering.getUuid()));
```
Comment (anuragaw) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: nit - vmInstance.name?

Comment (shwstppr) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: @anuragaw instead vmInstance.getUuid()? As we show UUIDs mostly.

**Comment by anuragaw** [comment]
LGTM

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -965,6 +965,9 @@ public UserVm upgradeVirtualMachine(UpgradeVMCmd cmd) throws ResourceAllocationE
         // Check resource limits for CPU and Memory.
         Map<String, String> customParameters = cmd.getDetails();
         ServiceOfferingVO newServiceOffering = _offeringDao.findById(svcOffId);
+        if (newServiceOffering.getState() == DiskOffering.State.Inactive) {
+            throw new InvalidParameterValueException(String.format("Unable to upgrade virtual machine %s with an inactive service offering %s", vmInstance.getUuid(), newServiceOffering.getUuid()));
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: 👍

**Comment by rhtyd** [comment]
@blueorangutan package

**Pull request merged**